File: explore/best-time-to-buy-and-sell-stock.go
package main

import "fmt"

func maxProfit(prices []int) int {
	var (
		maxP = 0
		buy  = 1<<31 - 1
	)
	for _, p := range prices {
		buy = min(buy, p)
		maxP = max(maxP, p-buy)
	}
	return maxP
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

func main() {
	fmt.Println(maxProfit([]int{1, 6, 3, 8, 4, 5, 9}))
}

File: 2018/explore/best-time-to-buy-and-sell-stock.py
class Solution:
	def maxProfit(self, prices):
		"""
		:type prices: List[int]
		:rtype: int
		"""
		buy_day = 0
		max_profit = 0
		for i in range(len(prices)):
			if prices[i] < prices[buy_day]:
				buy_day = i
			max_profit = max(max_profit, prices[i] - prices[buy_day])
		return max_profit
